### ISSUE TITLE
Add an easy way to share (non public) packages and templates

### ISSUE THREAD

**Issue by pachi** [typst/typst#4484]
### Description

Add basic package management to the typst cli to be able to add a package from a package directory or (ideally) a zipped package (.tpz?) and make it available to the @local namespace. The command line interface could be like this:

`  typst pkg install ./my-package-dir`  

or

`  typst pkg install ./my-package-bundle.tpz`  

This would ease the distribution of internal packages as well as package development, as it avoids having to know the details of where to copy the package (`  {cache-dir}/typst/packages/local/`   or an overridden location by env vars) and knowing the mapping between the package format and the package repository structure.

I suggested this feature previously in issue #3099 (https://github.com/typst/typst/issues/3099#issuecomment-2141873547 and https://github.com/typst/typst/issues/3099#issuecomment-2146864535 ) but that issue deals with other aspects of the @local namespace not directly related to this use case. That also included typst pkg new my-new-pkg-dir for basic scaffolding but this is orthogonal to this feature.

### Use Case

Many organizations have templates (and could use packages) that are not suitable for publication in the open due to privacy and other reasons. Also, package and template developers may want to only publish more or less finished versions of a package.

Currently, to use such templates or packages a user has the following options:

- copy the package dir along the Typst document using it  and do the imports with relative paths
- do a manual install of the package, to a platform specific location (could be env var specific) and use the @local namespace in imports
- install the third party typst package installer

The first one feels a bit clumsy and the second one is not very kind on normal users and need a lot of direction from the package/template author to get running.

The proposed solution would allow installing the package using 'typst pkg install your-new-fancy-pkg.tpz' and afterwards use '#import "@local/fancy-pkg:0.1.0": template' on your document.

**Comment by tingerrr** [comment]
As I've mentioned on the discord, you can already use [utpm](https://github.com/Thumuss/utpm/) for some of these tasks at the moment. But I agree that this should eventually be part of the CLI, including other package management related tasks like publishing and the like.

**Comment by pachi** [comment]
> As I've mentioned on the discord, you can already use [utpm](https://github.com/Thumuss/utpm/)
Yes, it's the third option, but a user would need to have a Rust toolchain, download utpm, compile it and start using that. I can imagine that a developer could do that, but that's too much for a normal user that just wants to craft their documents.

To me, the publishing features and even using git sources is much less important as that would affect package creators that are in some way already power users. Being able to make a template or package available to use locally without knowing the details of how Typst stores its data is much more important for any user. The .tpz is just a mere convenience to avoid unzipping, etc when distributing a package or template.

**Comment by tingerrr** [comment]
Yeah, I definitely agree, I think the issue with utpm at this moment is that there's no release action for people to easily install from, nor does it have all the features a new user would need.

I've tried to get more users to help contribute to make installation of packages easier with it, but so far this has not worked out, and I'm spread thin on many other projects.

I hope we'll end up with a good native cli interface for package management down the line. But I don't think it would take some random compressed archive, but rather the source forge at which it is located, such as GitHub.

**Comment by pachi** [comment]
Having a way to share packages and single files would still be useful, as some organizations may be very reluctant to share a public repo, such an official institution report template or internal mail templates with scanned signatures, etc, and, mainly, don't host git infrastructure.

**Comment by tingerrr** [comment]
I think hosting an on-premise registry is far more likely for a company.

**Comment by laurmaedje** [comment]
> I think hosting an on-premise registry is far more likely for a company.

And supported in the on-premises web app. :)

**Comment by jeremyforan** [comment]
Is there a major difference between the process that pulls from the '@preview' namespace and what would be the '@local' namespace. 

Apologies for my ignorance, wouldn't install a local package be similar to the process of caching a hosted package from the '@preview' namespace?

I like 'typst pkg install your-new-fancy-pkg.tpz' and would also like, 'typst pkg install https://github/jeremyforan/bf-branded-theme'.

**Comment by jeremyforan** [comment]
My primary use case is to be able to standardize a theme for my company. Staff should be able to complete focus on the content. If they find an issue with the formatting or would like to make a tweak to the layout, I can update the theme independent of the project. Once I commit the changes they can simply update the theme version. Over time the theme will evolve to be more dynamic and configurable, taking into account different document types.

**Comment by tingerrr** [comment]
> Is there a major difference between the process that pulls from the '@Preview' namespace and what would be the 'https://github.com/Local' namespace.
> 
> Apologies for my ignorance, wouldn't install a local package be similar to the process of caching a hosted package from the '@Preview' namespace?
> 
> I like 'typst pkg install your-new-fancy-pkg.tpz' and would also like, 'typst pkg install https://github/jeremyforan/bf-branded-theme'.

In general, it works like this: If a path with a package spec is requested from the compiler, the system interface (World) does the following:
1. Check if the package exists in the data directory, if yes return it.
2. Check if the package exists in the cache directory, if yes return it.
3. Check if its namespace is `preview`, if yes download it.

Any other namespace can just be created b users as they wish, and it will be found if it exists, preview is just the only one that supports built-in on-demand downloading.

Git packages are something that's planned, but I think it has been discussed that they'll likely get a declarative solution, i.e. declaring those dependencies in a `typst.toml` of the project using them.

Whether they are valid arguments to such an installation command is up to debate.

**Comment by jeremyforan** [comment]
Selfishly, it would be great if it worked like Go. I always enjoyed working with their package management.
```bash
go get
```
or
```bash
 go mod tidy
 ```
 
Assuming the packages you want to import are already declared in the files that will use them, than ideally having a file that is generated automatically and can be checked when needed would be great.

**Comment by JeppeKlitgaard** [comment]
In line with this, I think it could useful and important for Typst to have the ability easily and natively use packages that are not included in the central Typst Universe. A very common use-case in other languages package managers is having a dependency from a git repository. I think it would be very nice if the Typst import statement could support this natively, for example:

```typst
#import @git/git:github.com:User/Package.git
#import @git/git+https://github.com/User/Package.git
```

While there are definitely better examples, I would quite like to have a personal preamble/package that I host on GitHub, but would not necessarily be of a publishable quality.

Other 'handlers' could be added (`@web/zip+https://somesite.com/files/package.typst.zip`)

Perhaps this belongs in an issue of its own, but I thought it was relevant to add here as well

**Comment by laurmaedje** [comment]
@JeppeKlitgaard See https://github.com/typst/typst/issues/4040
